### DISCUSSION TITLE
Fabric: Peripheral making with non-block-entities?

### DISCUSSION THREAD

**Discussion by piprett** [cc-tweaked/CC-Tweaked#1359] [Q&A] [answered]
Hey. I was going to develop a peripheral mod, and I have been digging around with my GitHub code search, and I believe that it is not possible to create peripherals for blocks without blockentities on fabric, while it is on forge? I am assuming this based on #1208? The PeripheralLookup only appears to support block entities? Please correct me if I am wrong. I am posting it as a discussion, as I assume there is a way...

**Comment by SquidDev** [comment, marked as answer]
The `BlockLookupApi` (which is what `PeripheralLookup` uses under the hood) should allow registering both blocks (with `registerForBlocks`) and block entities (`registerForBlockEntity`) - see the [javadoc in the original source](https://github.com/FabricMC/fabric/blob/4626a20621d40be4ba227d8a6d2524012c25cf72/fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/api/lookup/v1/block/BlockApiLookup.java#L196-L222)

**Reply by piprett** [reply]
After what I can see the `registerForBlocks` expects a function with five arguments, `(world, pos, state, blockEntity, direction)`? But then what is this `blockEntity`? The docs I am getting this information from might be wrong I guess?

**Reply by SquidDev** [reply]
It's [the block entity which _might_ exist in the world](https://github.com/FabricMC/fabric/blob/4626a20621d40be4ba227d8a6d2524012c25cf72/fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/api/lookup/v1/block/BlockApiLookup.java#LL277C25-L277C69) - literally just the result of `world.getBlockEntity`. If your block doesn't have a block entity, then this will be `null`.

**Comment by blockninja124** [comment]
Is it possible to do this on forge? Capabilities seem to require a BlockEntity

**Reply by SquidDev** [reply]
Assuming you're on 1.20, you can register an [`IPeripheralProvider`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.20.x/projects/forge-api/src/main/java/dan200/computercraft/api/peripheral/IPeripheralProvider.java).
